engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/graph/ScopeImpl.scala

```diff
@@ -0,0 +1,377 @@
+package org.enso.compiler
+package pass.analyse
+package alias.graph
+
+import org.enso.compiler.core.CompilerError
+
+import scala.collection.immutable.HashMap
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+/** A representation of a local scope in Enso.
+  *
+  * @param childScopes all scopes that are _direct_ children of `this`
+  * @param occurrences all symbol occurrences in `this` scope indexed by the identifier of the name
+  * @param allDefinitions all definitions in this scope, including synthetic ones.
+  *                       Note that there may not be a link for all these definitions.
+  */
+sealed private[graph] class ScopeImpl(
+  private[graph] var _childScopes: List[ScopeImpl] = List(),
+  private[graph] var _occurrences: Map[GraphImpl.Id, GraphOccurrence] =
+    HashMap(),
+  _defs: java.util.List[
+    GraphOccurrence.Def
+  ] = new java.util.ArrayList()
+) extends Graph.Scope {
+  private[graph] val _allDefinitions: java.util.List[
+    GraphOccurrence.Def
+  ] = new java.util.ArrayList(_defs.stream.map(_.withScope(this)).toList)
+
+  private[graph] var _parent: ScopeImpl = null
+
+  def childScopes    = _childScopes
+  def occurrences    = _occurrences
+  def allDefinitions = java.util.Collections.unmodifiableList(_allDefinitions)
+  def parent: Option[ScopeImpl] =
+    if (this._parent eq null) None else Some(_parent)
+
+  /** Counts the number of scopes from this scope to the root.
+    *
+    * This count includes the root scope, but not the current scope.
+    *
+    * @return the number of scopes from this scope to the root
+    */
+  private[analyse] def scopesToRoot: Int = {
+    parent.flatMap(scope => Some(scope.scopesToRoot + 1)).getOrElse(0)
+  }
+
+  /** Sets the parent of the scope.
+    *
+    * The parent scope must not be redefined.
+    *
+    * @return this scope with parent scope set
+    */
+  final def withParent(parentScope: Graph.Scope): this.type = {
+    if (this._parent ne parentScope) {
+      org.enso.common.Asserts.assertInJvm(parent.isEmpty)
+      this._parent = parentScope.asInstanceOf[ScopeImpl]
+    }
+    this
+  }
+
+  /** Creates a structural copy of this scope, ensuring that replicated
```

Comment:
Outdated docs - now this method also mutates `mapping` parameter

Reply:
Mutating the `mapping` parameter was always the case. I'd like to get rid of such a coding pattern, but it seems to be needed. The usages do the `deepCopy` and then read the mapping of old to new `Scope` instances.